apps/server/src/shared/infra/mail/mail.service.ts

```diff
@@ -10,12 +11,53 @@ interface MailServiceOptions {
 
 @Injectable()
 export class MailService {
+	private readonly domainBlacklist: string[];
+
 	constructor(
 		private readonly amqpConnection: AmqpConnection,
-		@Inject('MAIL_SERVICE_OPTIONS') private readonly options: MailServiceOptions
-	) {}
+		@Inject('MAIL_SERVICE_OPTIONS') private readonly options: MailServiceOptions,
+		private readonly configService: ConfigService<IMailConfig, true>
+	) {
+		this.domainBlacklist = this.configService.get<string[]>('ADDITIONAL_BLACKLISTED_EMAIL_DOMAINS');
+	}
 
 	public async send(data: Mail): Promise<void> {
+		if (this.domainBlacklist.length > 0) {
+			data.recipients = this.removeEmailAddressesThatHaveBlacklistedDomain(
+				data.recipients,
+				this.domainBlacklist
+			) as string[];
+			data.cc = this.removeEmailAddressesThatHaveBlacklistedDomain(data.cc, this.domainBlacklist);
+			data.bcc = this.removeEmailAddressesThatHaveBlacklistedDomain(data.bcc, this.domainBlacklist);
+			data.replyTo = this.removeEmailAddressesThatHaveBlacklistedDomain(data.replyTo, this.domainBlacklist);
+		}
+
+		if (data.recipients.length === 0) {
+			return;
+		}
+
 		await this.amqpConnection.publish(this.options.exchange, this.options.routingKey, data, { persistent: true });
 	}
+
+	private removeEmailAddressesThatHaveBlacklistedDomain(
```

Comment:
The name is very descriptive which is good, but I'd suggest using a shorter names in the future - they do not always have to be very precise - I think something like even just `filterEmailAddresses()` would be sufficient as it's easy to understand what it does in that particular case (as it's a class method, not some shared function).

But you know, as Phil Karlton said, _"There are only two hard things in Computer Science: cache invalidation and naming things."_ so don't worry, it's not easy in general 😊 .

Reply:
changed